src/service/fixed-layer.js

```diff
@@ -0,0 +1,417 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {onDocumentReady} from '../document-state';
+import {log} from '../log';
+import {setStyle, setStyles} from '../style';
+
+const TAG = 'FixedLayer';
+
+
+/**
+ * The fixed layer is a *sibling* of the body element. I.e. it's a direct
+ * child of documentElement. It's used to manage the `postition:fixed`
+ * elements in iOS-iframe case due to the
+ * https://bugs.webkit.org/show_bug.cgi?id=154399 bug, which is itself
+ * a result of workaround for the issue where scrolling is not supported
+ * in iframes (https://bugs.webkit.org/show_bug.cgi?id=149264).
+ * This implementation finds all elements that could be `fixed` and checks
+ * on major relayouts if they are indeed `fixed`. All `fixed` elements are
+ * moved into the fixed layer.
+ */
+export class FixedLayer {
+  /**
+   * @param {!Document} doc
+   * @param {!Vsync} vsync
+   * @param {number} paddingTop
+   * @param {boolean} transfer
+   */
+  constructor(doc, vsync, paddingTop, transfer) {
+    /** @const {!Document} */
+    this.doc = doc;
+
+    /** @private @const {!Vsync} */
+    this.vsync_ = vsync;
+
+    /** @private {number} */
+    this.paddingTop_ = paddingTop;
+
+    /** @private @const {boolean} */
+    this.transfer_ = transfer;
+
+    /** @private {?Element} */
+    this.fixedLayer_ = null;
+
+    /** @const @private {!Array<!FixedElementDef>} */
+    this.fixedElements_ = [];
+  }
+
+  /**
+   * @param {boolean} visible
+   */
+  setVisible(visible) {
+    if (this.fixedLayer_) {
+      this.vsync_.mutate(() => {
+        setStyle(this.fixedLayer_, 'visibility',
+            visible ? 'visible' : 'hidden');
+      });
+    }
+  }
+
+  /**
+   * Must be always called after DOMReady.
+   */
+  setup() {
+    // Find all `position:fixed` elements.
+    const stylesheets = document.styleSheets;
+    const fixedSelectors = [];
+    for (let i = 0; i < stylesheets.length; i++) {
+      const stylesheet = stylesheets[i];
+      if (stylesheet.disabled ||
+              !stylesheet.ownerNode ||
+              stylesheet.ownerNode.tagName != 'STYLE' ||
+              stylesheet.ownerNode.hasAttribute('amp-boilerplate') ||
+              stylesheet.ownerNode.hasAttribute('amp-runtime') ||
+              stylesheet.ownerNode.hasAttribute('amp-extension')) {
+        continue;
+      }
+      this.discoverFixedSelectors_(stylesheet.cssRules, fixedSelectors);
+    }
+
+    try {
+      const idMap = {};
+      fixedSelectors.forEach(selector => {
+        const elements = this.doc.querySelectorAll(selector);
+        for (let i = 0; i < elements.length; i++) {
+          if (i > 10) {
+            // We shouldn't have too many of `fixed` elements.
```

Comment:
Not yet. We are still trying to figure out the best way to enable dev. But otherwise, this is a very generous constraint.